### DISCUSSION TITLE
How to restrict a generic class to a Protocol?

### DISCUSSION THREAD

**Discussion by jugaadi** [python/typing#1316] [Q&A]
```python
from typing import Protocol

class A(Protocol):
    def x(): ...
    def y(): ...

class B:
    def __getattr__(self, name) -> Callable:
    

b = B()
b.x()
b.y()
b.z()  # error
```

Is there any way to limit B from accepting attributes not compatible with `A` protocol?

**Comment by erictraut** [comment]
Can you provide more details? What do you mean by "restrict B"? Also, `A` is not currently defined as a protocol in your example, Did you intend for `A` to be a subclass of `Protocol`?

**Reply by jugaadi** [reply]
@erictraut 

Sorry.. Ive updated the message with the corrections and an example. Since any method can be called via B, I wanted to know whether we can constrain the attributes that can be called from B by associating it with A.

To achieve the same result, Im currently using the below solution:

> def \_\_getattr\_\_(self, name: Literal['x', 'y']):

**Reply by erictraut** [reply]
If you want to ensure that `B` provides the methods required for protocol `A`, I recommend not supplying a `__getattr__` method and instead defining individual methods `x` and `y` within `B`. If `B` subclasses from `A`, type checkers will validate it defines all required methods for `A`.
